### PR TITLE
refactor!: add processor builder and transform buidler

### PR DESCRIPTION
**Pull request by paomian**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

* remove dissect reference keys. because the resulting key is uncertain, it can't be specified by transform; if it is certain, you can use name matching to get the specified key directly.
* make all processors modify the intermediate state of the vec format directly. optimizing performance
* use the builder to build all processors, making the processor build flow easier to understand.

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [x] This PR requires documentation updates.

### CODE REVIEW

File: src/pipeline/benches/processor.rs [hunk 1]
```diff
@@ -249,11 +235,13 @@ fn criterion_benchmark(c: &mut Criterion) {
...
-    group.bench_function("processor map", |b| {
-        b.iter(|| processor_map(black_box(&pipeline), black_box(input_value.clone())))
-    });
     group.bench_function("processor mut", |b| {
-        b.iter(|| processor_mut(black_box(&pipeline), black_box(input_value.clone())))
+        b.iter(|| {
+            let result = processor_mut(black_box(&pipeline), black_box(input_value.clone()));
+            if result.is_err() {
+                panic!("Error: {:?}", result);
+            }
```
Comment (sunng87) [inline src/pipeline/benches/processor.rs:243]:
```suggestion
            processor_mut(black_box(&pipeline), black_box(input_value.clone())).unwrap();
```

File: src/pipeline/src/etl/transform.rs [hunk 2]
```diff
@@ -222,31 +266,31 @@ impl Default for Transform {
...
+    //         (_, Value::Null) => Ok(()), // if default is not set, then it will be regarded as default null
+    //         (_, _) => {
+    //             let target = self
+    //                 .type_
+    //                 .parse_str_value(default.to_str_value().as_str())?;
+    //             self.default = Some(target);
+    //             Ok(())
+    //         }
+    //     }
+    // }
```
Comment (sunng87) [inline src/pipeline/src/etl/transform.rs:293]: remove  this if not needed any more

File: src/pipeline/src/etl/transform.rs [hunk 3]
```diff
@@ -222,31 +266,31 @@ impl Default for Transform {
...
 
 impl Transform {
-    fn with_fields(&mut self, mut fields: Fields) {
-        update_one_one_output_keys(&mut fields);
-        self.fields = fields;
-    }
+    // fn with_fields(&mut self, mut fields: Fields) {
+    //     update_one_one_output_keys(&mut fields);
+    //     self.fields = fields;
+    // }
```
Comment (sunng87) [inline src/pipeline/src/etl/transform.rs:272]: remove if not needed

**Review by sunng87 — APPROVED** [review]
This is hard to review. I'm +1 to move this forward. But there are some points we need to verify and refactor in future patches:

- make sure there is no `unwrap` in production code
- don't use `String` as `Err` in pipeline `Result`. Follow the `Error` processing pattern.
- make sure no existing test is modified except the breaking change. We need to keep the behaviour correct

**Comment by paomian** [comment]
> This is hard to review. I'm +1 to move this forward. But there are some points we need to verify and refactor in future patches:
> 
> * make sure there is no `unwrap` in production code
> * don't use `String` as `Err` in pipeline `Result`. Follow the `Error` processing pattern.
> * make sure no existing test is modified except the breaking change. We need to keep the behaviour correct

It's no unwrap in production code.
We will replace the String error type.
I deleted some of the tests because of some breaking changes, but the original tests all passed.

**Review by shuiyisong — APPROVED** [review]
Let's move this pr forward. Remember to update docs for the breaking change.

**Pull request merged**